include/ghex/common/await_futures.hpp

```diff
@@ -12,75 +12,71 @@
+#include <algorithm>
 
 namespace gridtools {
+namespace ghex {
 
-    namespace ghex {
-        
-        /** @brief wait for all futures in a range to finish and call 
-          * a continuation with the future's value as argument. */
-        template<typename Future, typename Continuation>
-        void await_futures(std::vector<Future>& range, Continuation&& cont)
-        {
-            int size = range.size();
-            while(size>0)
-            {
-                for (int i=0; i<size; ++i)
-                {
-                    if (range[i].test())
+/** @brief wait for all futures in a range to finish and call 
+  * a continuation with the future's value as argument. */
+template<typename Future, typename Continuation>
+inline void await_futures(std::vector<Future>& range, Continuation&& cont)
+{
+    static thread_local std::vector<int> index_list;
+    index_list.resize(range.size());
+    std::iota(index_list.begin(), index_list.end(), 0);
+    while (index_list.size())
+    {
+        index_list.resize(
+                std::remove_if(index_list.begin(), index_list.end(),
+                    [&range, cont = std::forward<Continuation>(cont)](int idx)
                     {
-                        cont(range[i].get());
-                        --size;
-                        if (i<size)
-                            range[i--] = std::move(range[size]);
-                        range.pop_back();
-                    }
-                }
-            }
-        }
-        
-        template<typename Future>
-        void await_futures(std::vector<Future>& range)
-        {
-            int size = range.size();
-            while(size>0)
-            {
-                for (int i=0; i<size; ++i)
-                {
-                    if (range[i].test())
-                    {
-                        --size;
-                        if (i<size)
-                            range[i--] = std::move(range[size]);
-                        range.pop_back();
-                    }
-                }
-            }
-        }
-        
-        template<typename Communicator, typename Request>
-        void await_requests(Communicator comm, std::vector<Request>& range)
-        {
-            int size = range.size();
-            while(size>0)
-            {
-                comm.progress();
-                for (int i=0; i<size; ++i)
-                {
-                    if (range[i].test())
-                    {
-                        --size;
-                        if (i<size)
-                            range[i--] = std::move(range[size]);
-                        range.pop_back();
-                    }
-                }
-            }
-        }
+                        if (range[idx].test())
+                        {
+                            cont(range[idx].get());
+                            return true;
+                        } else return false;
+                    })
+                - index_list.begin());
```

Comment:
why do you need to resize all the time?

Reply:
the `index_list` is a helper data structure - it needs to be the same size as the vector of futures/request. Could you elaborate on your concerns here? The reason I made these helper vectors `static thread_local` is to avoid frequent reallocation (the `resize` will not trigger an allocation if the capacity is big enough).

Reply:
it's probably not a problem, but I would have done in in two phases: one I compute the number of things to remove and then I resize. Which is probably not better, since the resize if a decrement of a variable...

Reply:
ok, now I understood. Good point, have changed it accordingly.

Reply:
ok, now I understood. Good point, have changed it accordingly.